USB: Remote NDIS USB Ethernet support

finikorg opened the pull request:
Add RNDIS protocol, WIP.

finikorg committed:
usb: rndis: Add Remote NDIS protocol handling

Add implementation of Microsoft Remote NDIS USB Ethernet protocol.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: rndis: Add RNDIS device descriptors to the table

Add needed descriptors for USB Remote NDIS Ethernet Network.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: zperf: Add USB RNDIS configuration check

Add configuration for zperf sample allowing to build USB dongle with
zperf for testing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: Correct default SYS_LOG_USB and SYS_LOG_USB_DRIVER levels

Making configuration files useful.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Rename SYS_LOG_USB to SYS_LOG_USB_DEVICE

At the moment CONFIG_SYS_LOG_USB_LEVEL name does not specify that this
is log level for the Device Stack. Make it clear renaming to the
proper name.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Use standard SYS_LOG_USB_DEVICE_LEVEL for log level

Change hardcoded value to configuration defined.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add missing definitions needed for RNDIS

Add missing CDC definitions for RNDIS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add init() callback to network device functions

Add init() callback needed for RNDIS and maybe others functions.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: netusb: Remove unneeded debug

Remove old unneeded debug statements.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@loicpoulain

loicpoulain commented:
General notes since I have EEM in the pipe as well:
- We systematically use an extra buffer to send data to usb, usually linerarization in a ep buffer. Maybe we could avoid this by making usb subsystem handling net_buf directly.
- In my case I need to send multi-packet in a same transfer which does work today with stm32 driver only, we should align both driver. e.g. sending a packet greater than ep packet size is split in multiple packets but sent during the same usb transfer, this improves throughput and reduces latency. Don't know if RNDIS support/request this btw.

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	}
+
+	len = sys_le32_to_cpu(hdr->len);
+
```
you can remove that empty line, your if condition is testing the previous assignment.

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	    /*
+	     * Calculate additional offset since payload_offset is calculated
+	     * from the start of itself ;)
+	     */
```
Ouch this comment in the middle of the if condition makes it hard to read. move it before the if

finikorg replied:
This comment is related to part of the "if" condition, if I move it before we would not know where it relates to.

tbursztyka replied:
You could explain the whole if then. It's quite ugly to cut the condition in the middle

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	}
+}
+
+struct tlv {
```
declare that structure at the beginning of the file.

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+}
+
+static int rndis_class_handler(struct usb_setup_packet *setup, s32_t *len,
+			       u8_t **data)
```
wrong indent here it seems.

finikorg replied:
Looks OK in text editor

tbursztyka replied:
ok so it's github messing up again.

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+		 */
+		queue_encapsulated_cmd(*data, *len);
+	} else if (setup->bRequest == CDC_GET_ENC_RSP &&
+		 REQTYPE_GET_DIR(setup->bmRequestType) == REQTYPE_DIR_TO_HOST) {
```
same

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+				try_write(CONFIG_RNDIS_IN_EP_ADDR, zero,
+					  sizeof(zero));
+			}
+		}
```
you could move this code block to its own function.

tbursztyka commented on include/usb/class/usb_cdc.h:
```diff
@@ -65,6 +65,8 @@
  * @note PSTN120.pdf, 6.3, Table 13
  */
+#define CDC_SEND_ENC_CMD		0x00
+#define CDC_GET_ENC_RSP			0x01
```
Is github messing up the order of the commits again? I guess this commit comes _before_ adding RNDIS

finikorg replied:
Yes, it comes before.

tbursztyka commented on subsys/usb/class/netusb/Kconfig:
```diff
@@ -78,4 +78,41 @@ config USB_DEVICE_NETWORK_ECM_MAC
 
 endif # USB_DEVICE_NETWORK_ECM
 
+if USB_DEVICE_NETWORK_RNDIS
```
Merge that patch with "usb: rndis: Add Remote NDIS protocol handling"

finikorg replied:
OK

tbursztyka reviewed (changes requested):
Tiny comments. Overall looks fine to me

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	u32_t rx_no_buf;
+
+	atomic_t notify_count;
+} rndis = {
```
Is there any possibility this struct can be accessed from different contexts, such as different threads or a interrupt context? If so, this must be protected.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	.skip_bytes = 0,
+};
+
+static u8_t manufacturer[] = { 't', 'e', 's', 't' };
```
Is "test" an acceptable hardcoded value?  Can this be marked as `const` as well, since I don't see anything modifying it?

finikorg replied:
This is my left-overs from the old code base, now we use special CONFIG_USB_DEVICE_MANUFACTURER, I will fix it.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	     * Calculate additional offset since payload_offset is calculated
+	     * from the start of itself ;)
+	     */
+	    (u8_t *)&hdr->payload_offset - (u8_t *)hdr) {
```
If I'm understanding correctly, `payload_offset` came directly from the wire and it has not been properly validated.  This could lead to problems if the header is malformed, maliciously or not.

Also, the type of a pointer difference is `ptrdiff_t`, a signed integral type; len is an unsigned integral type; will this work in all occasions?

finikorg replied:
NO! This is offset of the payload_offset field from the start of the structure hdr.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+{
+
+	if (!k_fifo_is_empty(&rndis_tx_queue)) {
+#if 0
```
Please remove commented-out code

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+struct tlv {
+	u32_t type;
+	u32_t len;
+	u8_t data[0];
```
Do we need to support *very* old GCC compilers? Might be better to just declare `data` as `u8_t data[]`, that is standard C99.

finikorg replied:
ok

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	enum {
+		UNINITIALIZED,
+		INITIALIZED,
+	} state;
```
Will there be another state? Can this be a `bool initialized` instead?

finikorg replied:
at the moment I have 2 states, maybe there will be more in the future

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+static K_THREAD_STACK_DEFINE(cmd_stack, 1024);
+static struct k_thread cmd_thread_data;
+
+static struct __rndis {
```
This struct can be better packed by reorganizing its members so that there are no holes due to alignment.

finikorg replied:
I am not sure we need to organize members of the structures by alignment rather then by logical order.

jukkar replied:
I agree with @lpereira here, we should align structs for minimizing memory consumption. The ordering of items might look funny but that is not very relevant. You could add more comments into this struct to explain more what the various integers like net_filter, ep_in, ep_notify are in this struct.

finikorg replied:
Actually I will remove most of fields since we have now new interface for netusb.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+
+	SYS_LOG_DBG("EP 0x%x status %d len %u", ep, ep_status, len);
+
+	usb_read(ep, buffer, len, NULL);
```
Is there any possibility of `len` being larger than CONFIG_RNDIS_BULK_EP_MPS?

finikorg replied:
It shouldn't but I have added check and limit read to MPS.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+		hdr_offset = sizeof(struct rndis_payload_packet);
+	}
+
+	ret = net_pkt_append_all(rndis.in_pkt, len - hdr_offset,
```
* `hdr_offset` should be declared as an `u32_t` to avoid converting it from `u8_t` here
* attributing to it the result of the `sizeof()` operator will truncate `size_t` to `u8_t` (as the code is written now); is this OK?
* Can `len` be smaller than `hdr_offset` in a way that the unsigned subtraction will lead to overflow?

finikorg replied:
I have changed hdr_offset to u32_t though it cannot be so large since this is offset inside 64 bytes array.
hdr_offset is not zero if we get a new packet which starts with header of this offset size. In parse_rndis_header() we already have a check
	if (buf_len < sizeof(*hdr)) {
		SYS_LOG_ERR("Too small packet len %u", buf_len);
		return -EINVAL;
	}

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	}
+
+	ret = net_pkt_append_all(rndis.in_pkt, len - hdr_offset,
+				 buffer + hdr_offset, K_FOREVER);
```
Is `buffer + hdr_offset` guaranteed to be within the boundaries of `buffer`?

finikorg replied:
Yes, hdr_offset is not zero only for new packet with header

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+	rsp->max_packets = sys_cpu_to_le32(1);
+	rsp->max_transfer_size = sys_cpu_to_le32(rndis.mtu +
+						 sizeof(struct net_eth_hdr) +
+						 44 + 22);
```
What are these magic numbers?

finikorg replied:
Fixed

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+		break;
+	case RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE");
+		net_buf_add_le32(buf, 1558);
```
What is this magic value?

finikorg replied:
This is maximum total packet length 
https://docs.microsoft.com/en-us/windows-hardware/drivers/network/oid-gen-maximum-total-size

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+static u8_t *rndis_hdr_add(u8_t *buf, u32_t len)
+{
+	struct rndis_payload_packet *hdr = (void *)buf;
+	u8_t offset = (u8_t *)&hdr->payload_offset - (u8_t *)hdr;
```
Beware of casting `ptrdiff_t` to `u8_t`.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+
+static int rndis_send(struct net_pkt *pkt)
+{
+	u8_t buf[1500], *p; /* FIXME */
```
Shouldn't 1500 be a named constant that's used also by the rndis struct?

finikorg replied:
Fixed, I am planning also to refactor send() function similar to ECM one.

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+		       net_pkt_ll_reserve(pkt) + pkt->frags->len);
+		p += net_pkt_ll_reserve(pkt) + pkt->frags->len;
+
+		for (frag = pkt->frags->frags; frag; frag = frag->frags) {
```
A chain of fragments can easily overflow `buf` here.

lpereira replied:
This looks like something that still needs to be addressed.

lpereira commented:
I haven't finished reviewing this, but I'm sure I'll find more things if I keep poking around.  Please review validation, pointer arithmetic, and type truncation (specially when pointer arithmetic is used; you may want to build with `-Wconversion` to catch these errors).

tbursztyka commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1000 @@
+
+	usb_read(ep, buffer, len, NULL);
+
+	/* We already keep frame keeping with len, warn here about
```
this comment would need to be rephrased.

finikorg replied:
Fixed

finikorg committed:
usb: Set Class Handler buffer size to 256 for RNDIS

Set default value for RNDIS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: rndis: Select USB_COMPOSITE_DEVICE for RNDIS

Windows works better with this option.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
I addresses most of the comments, still need to address couple of issues, @lpereira how can I pass now with cmake -W parameters? I tried "cd outdir/<board> && make -Wconversion", no warnings ;)

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1018 @@
+	return buf + sizeof(*hdr);
+}
+
+static u8_t buf[RNDIS_GEN_MAX_TOTAL_SIZE];  /* FIXME */
```
`buf` is not a good name for a global variable; it's in fact very easy to shadow it locally.  Please be more specific: `send_buffer` is preferable.

Also, is there a possibility that two threads will call `rndis_send()`?

In addition: what's that FIXME comment is referring to?

finikorg replied:
About "(u8_t *)&hdr->payload_offset - (u8_t *)hdr)" construction: those are pointers to struct hdr, they should not be validated!
Now for some reason I cannot reply to all github comments separately. So how net_pkt_append_all() can be called if in_pkt == NULL?
For other comments I am rewriting send() function so those issues would be fixed.

finikorg replied:
I also have a check: 
if (buf_len < sizeof(*hdr)) {
		SYS_LOG_ERR("Too small packet len %u", buf_len);
		return -EINVAL;
	}

lpereira commented:
BTW, you should be able to pass CFLAGS as environment variables to CMake: `CFLAGS="-Wconversion -Wall -Wextra -Werror" cmake ..`.

finikorg committed:
usb: netusb: Add Networking logging debug level

Adding Networking USB Device Debug level allows to fine-tune
logging.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
Updated rndis_send() function and other fixes

galak commented:
@lpereira are you ok with the changes?

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1054 @@
+	rsp->req_id = cmd->req_id; /* same endianness */
+
+	/* Parameter starts at the beginning of req_id field ;) */
+	param = (u8_t *)&cmd->req_id + sys_le32_to_cpu(cmd->buf_offset);
```
`cmd->buf_offset` also comes directly from the USB peer and it's used as an offset here.  This allows arbitrary memory reads.

finikorg replied:
I will fix code to add check that param is not going out of the range in this case.

lpereira reviewed (changes requested):
There are still problems with this patch that could lead to arbitrary memory access.  I didn't go through the whole driver, but I'm sure there are more similar issues.  Please validate all values used as offsets and pointers when they're received from the USB peer.

finikorg commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1048 @@
+
+static int parse_rndis_header(const u8_t *buffer, u32_t buf_len)
+{
+	struct rndis_payload_packet *hdr = (void *)buffer;
```
@lpereira I think payload_offset is validated in the following sentences. It is unsigned and we only validate that it is not too big.

finikorg replied:
@lpereira Could you check the latest code?

finikorg committed:
usb: Add RNDIS packet boundary check

Add check for rndis_set_handle() packet.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@lpereira have you had time to review the changes?

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1070 @@
+	 */
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
+	    sys_le32_to_cpu(hdr->payload_len) +
+	    (u8_t *)&hdr->payload_offset - (u8_t *)hdr) {
```
To calculate this expression (`(u8_t *)&hdr->payload_offset - (u8_t *)hdr`), it's better to use `offsetof()`, as it's performed in compile time.

finikorg replied:
I will use this macro

lpereira commented on subsys/usb/class/netusb/function_rndis.c:
```diff
@@ -0,0 +1,1070 @@
+	 * Calculate additional offset since payload_offset is calculated
+	 * from the start of itself ;)
+	 */
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
```
Would there be any issue if `payload_offset + payload_len - offsetof(header, payload_offset)` overflowed?  Unsigned overflow would mean that this would wrap around, creating a value that is smaller than `len`.

finikorg replied:
We get from this function only len, of course USB device could put anything to any field and we cannot always verify it, so in a case packet is broken it would be dropped by network interface.

finikorg replied:
@lpereira what do you think?

finikorg committed:
usb: rndis: Use offsetof macro to ease offset calculations

Use macro offsetof() for calculating offset inside structure.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

lpereira commented:
I'm approving the changes to move this forward; there might be other issues lingering around, though.

jli157 commented:
@finikorg How have you done the tests on Windows? I found the latest Windows 10 has problem to support the device with this driver enabled. I can't manually set the IP address on the host side. Should I report an issue for it given that the RNDIS driver is supposed to be well supported by Windows?

finikorg commented:
@jli157 was working in Windows when I used it last time, depending on Windows version there might be issues of selecting right driver, IIRC "Compatible RNDIS adapter, etc". With MS OS Descriptors driver should be selected automatically.

jli157 commented:
> @jli157 was working in Windows when I used it last time, depending on Windows version there might be issues of selecting right driver, IIRC "Compatible RNDIS adapter, etc". With MS OS Descriptors driver should be selected automatically.

Windows 10 seems not to be able to find the driver automatically. I have to manually update it to either "compatible RNDIS adapter" or "USB RNDIS Adapter". However, the IP address on the host can't be changed on either one.

jli157 commented:
While using the driver with Ubuntu 18.04, I often found the lots of errors like below:

```
[00:00:21.063,568] <err> usb_rndis: Not enough pkt buffers, len 118, skip 54
[00:00:21.073,822] <wrn> usb_rndis: Skip 54 bytes out of remaining 54 bytes
[00:00:21.084,014] <err> usb_rndis: Not enough pkt buffers, len 615, skip 551
[00:00:21.094,360] <wrn> usb_rndis: Skip 64 bytes out of remaining 551 bytes
[00:00:21.104,614] <wrn> usb_rndis: Skip 64 bytes out of remaining 487 bytes
```
 
And these:
```
[00:04:01.442,443] <err> usb_rndis: Error parsing RNDIS header
[00:04:01.451,477] <err> usb_rndis: Wrong data packet type 0x785bc0ef
[00:04:01.460,937] <err> usb_rndis: Error parsing RNDIS header
[00:04:01.469,909] <err> usb_rndis: Too small packet len 21
[00:04:01.478,515] <err> usb_rndis: Error parsing RNDIS header
[00:04:15.355,102] <err> usb_rndis: Error writing data to pkt: 0x2001f30c
[00:04:15.365,081] <err> usb_rndis: Wrong data packet type 0x11ff0040
[00:04:15.374,572] <err> usb_rndis: Error parsing RNDIS header

```

Any suggestions to fix these issues, increasing the RX buffer size?
